image/image.go

```diff
@@ -153,38 +172,105 @@ func sanitize(s string) string {
 	return strings.Replace(s, "/", "-", -1)
 }
 
-// getSchema1Manifest parses the manifest into a data structure, cleans it up, and returns it.
-// NOTE: The manifest may have been modified in the process; DO NOT reserialize and store the return value
-// if you want to preserve the original manifest; use the blob returned by Manifest() directly.
-// NOTE: It is essential for signature verification that the object is computed from the same manifest which is returned by Manifest().
-func (i *genericImage) getSchema1Manifest() (genericManifest, error) {
-	manblob, err := i.Manifest()
+// compile-time check that manifestSchema2 implements genericManifest
+var _ genericManifest = (*manifestSchema2)(nil)
+
+type manifestSchema2 struct {
+	src               types.ImageSource
+	ConfigDescriptor  descriptor   `json:"config"`
+	LayersDescriptors []descriptor `json:"layers"`
+}
+
+type descriptor struct {
+	MediaType string `json:"mediaType"`
+	Size      int64  `json:"size"`
+	Digest    string `json:"digest"`
+}
+
+func (m *manifestSchema2) String() string {
+	return ""
+}
+
+func (m *manifestSchema2) Layers() []string {
+	blobs := []string{}
+	for _, layer := range m.LayersDescriptors {
+		blobs = append(blobs, layer.Digest)
+	}
+	return blobs
+}
```

Comment:
mmm this is tricky because the "Config" digest isn't in the _layers_ array in a manifest - it's on it's own in both schema1 and schema2

afaict the config has been never copied for v2s2 (and we never used v2s2)